_posts/2023-01-14-trivially-relocatable.md

```diff
@@ -0,0 +1,656 @@
+a non-zero data size in `sus::mem::relocate_by_memcpy`.
+
+#### Volatile and trivially relocatable
+
+Popping the stack, we were walking through our implementation of `sus::mem::relocate_by_memcpy`.
+The first condition checked is if the type is `volatile`. A `volatile` type can not be copied
+byte-by-byte with `memcpy()` without [introducing the chance of tearing](
+https://quuxplusone.github.io/blog/2018/07/13/trivially-copyable-corner-cases/). So these types
+are strictly excluded. This also helps with defining trivial relocatable for classes with volatile
+members as we will see.
+
+### Opting into trivially relocatable
+
+In Clang, a type may opt into being trivially relocatable with the class annotation
+`[[clang::trivial_abi]]`. However there are two important things missing from this attribute:
+
+1. The attribute only works on Clang. A standard library should work well across all compilers.
+1. The attribute can't take template parameters into account.
+
+While `[[clang::trivial_abi]]` works for `std::unique_ptr<T>`, it does so because the data member of
+the class is `T*` and a pointer is always trivially relocatable. A simple example where we can not
+use the attribute would be:
+
+```cpp
+template <class T>
+struct S {
+    T t;
+};
+```
+
+Here the struct `S` is trivially relocatable if `T` is. But marking the type with
+`[[clang::trivial_abi]]` would introduce Undefined Behaviour when `T` is not. And this is where the
+proposal [P1144R6](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p1144r6.html) does
+something important by adding a boolean expression to the `[[trivially_relocatable]]` attribute.
+This makes it possible to conditionally apply the attribute to different template instantiations of
+a template type when opting into being trivially relocatable explicitly.
+
+However, the proposal also automatically infers trivially relocatable from its members, so we
+wouldn't need the attribute at all in the above example. We need a more complex example to see when
+this matters:
+
+```cpp
+template <class T>
+struct S : InheritMyMoveOperation<T>, InheritMyDestructor<T> {
+};
+```
+
+Here neither `InheritMyMoveOperator` nor `InheritMyDestructor` have enough information to
+determine if the type `S` should be trivially relocatable. The member of type `T` can only be present
+in one of the two, and that would dictate the behaviour for `S` if its simply inferred. The
+ability for `S` to be trivially relocated must be determined based on the interaction of the move
+operation and destructor, of which only the author of `S` has the full picture. The same follows for
+members that provide implementation details and which may not define their own move operations or
+destructors, such as unions.
+
+In this example `S` is trivially relocatable, since the destructor is a no-op after being moved
+from. But we can't tell from its members directly. The author needs to vouch for this by opting

```

Comment:
It might be nice to phrase this positively instead of negatively, e.g. "However, this is something that can only be determined by manually reading the implementation; the compiler cannot directly determine this based on the types of its fields."

Reply:
Yeah, thanks!